### PR TITLE
feat: add Implementation Guides compiler script 

### PR DESCRIPTION
**Pull request by justinusmenzel**
Issue #, if available: INT-931

Description of changes:
Added a IGCompiler TypeScript class to src and a compile-igs.ts script to the scripts folder.  
Its purpose of the class is 
 * to search through a directory and find any sub directory containing an IG package.
 * collect high level name and version info for each IG package
 * make sure there aren't any missing dependencies 
 * make sure there are no circular dependencies
 * extract all the SearchParameters (needs to be parameterized in the future)
 * pass list of SearchParameter instances to SearchImplementationGuides.compile function
 * store output in a JSON file

The compile-igs.ts script does
 * parse command line arguments
 * import ImplementationGuides implementation (implementations in the future
 * invoke IGCompiler
 * catch and report errors

Checklist:

* [ ] Have you successfully deployed to an AWS account with your changes?
* [x] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: scripts/compile-igs.js
Comment (carvantes) [inline scripts/compile-igs.js]: Errors should be handled here, otherwise you get:
```
 UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch().
```


It's probably fine to just catch them all and `console.err` them

File: scripts/compile-igs.js
Comment (carvantes) [inline scripts/compile-igs.js]: I think it's ok to have this as an argument for flexibility, but i think that we should make it optional and default to `igPath='../implementationGuides/'` if no argument is provided. IMO most people will use the default location for simplicity

File: scripts/compile-igs.js
Comment (carvantes) [inline scripts/compile-igs.js]: We need to make an exception for `hl7.fhir.r4.core`. There's no need to compile the base FHIR definitions since they are already already bundled in FWoA.

File: scripts/compile-igs.js
Comment (carvantes) [inline scripts/compile-igs.js]: We need to rearrange this a little bit. `compile` should be called only once with an array of all the search params from all the IGs that must be compiled. My apologies if this wasn't clear before.

Also, the `compile` output must be saved under `./compiledImplementationGuides` as a json file.

**Review by carvantes — CHANGES_REQUESTED** [review]

**Comment by carvantes** [comment]
I updated the PR to merge into `develop` instead of `mainline`. That's the usual procedure for PRs in this repo. Merges from `develop` to `mainline` occur automatically after automated integ tests pass.

The branch switch brought a lot of merge commits into the PR. Looks like this occurred since you branched off from `mainline` instead of from `develop`. It can be cleaned up by doing the following:

1. branch off from development
1. make your changes
1. force push to `INT-931_load_and_parse_IGs` to fix this PR

I'd suggest to clean this up before doing further changes.

Please read our contributing guidelines for more details: https://github.com/awslabs/fhir-works-on-aws-deployment/blob/mainline/CONTRIBUTING.md#contributing-via-pull-requests

**Comment by carvantes** [comment]
FYI, I submitted a PR with the implementation of the compile method. https://github.com/awslabs/fhir-works-on-aws-search-es/pull/38

**Comment by justinusmenzel** [comment]
Thank you @carvantes for the great feedback. I just noticed the issue with calling compile for each resource instead of only once with a list of all the resources. 

Question: would it make sense for me to take the original code that I wrote in TypeScript and put it in a ts file and then let the script be basically just a one liner and just call the TS class. That way most of the code is in TS and I can add a test.

**Comment by carvantes** [comment]
> Thank you @carvantes for the great feedback. I just noticed the issue with calling compile for each resource instead of only once with a list of all the resources.
> 
> Question: would it make sense for me to take the original code that I wrote in TypeScript and put it in a ts file and then let the script be basically just a one liner and just call the TS class. That way most of the code is in TS and I can add a test.

That sounds good.

I did something similar on https://github.com/awslabs/fhir-works-on-aws-search-es/pull/38. `scripts/searchParamsCompiler/run.ts` is a small script that imports `SearchImplementationGuides.compile`

I'd suggest to have the small script in typescript as well and run it using [ts-node](https://www.npmjs.com/package/ts-node). It is possible to have it in raw javascript, but that'd require to import typescript-compiled files from `/lib` and to ensure that `tsc` runs before the script and that setup is a bit clunky.

File: scripts/compile-igs.js
Comment (justinusmenzel) [reply scripts/compile-igs.js]: Just one JSON file? Should the file be `./compiledImplementationGuides.json` or a JSON file in the `./compiledImplementationGuides` folder. If the latter then what filename should it be? 
`./compiledImplementationGuides/searchParameters.json` maybe, since we are just compiling the `SearchParameter` resources for now. 
Should then different type of resources compiled by different compiles and stored in different files?

**Comment by justinusmenzel** [comment]
I just looked at the `develop` branch and it says there ` This branch is 1 commit ahead, 73 commits behind mainline. ` 
73 commits behind mainline seems like a lot to me. Is that expected?

File: scripts/compile-igs.js
Comment (carvantes) [reply scripts/compile-igs.js]: `./compiledImplementationGuides/searchParameters.json` would work. We need a different file per compiler. Actually, something like `./compiledImplementationGuides/fhir-works-on-aws-search-es.json` is best. Naming it after the module that compiled it is better since future compilers may consume several resource types at once.

> Should then different type of resources compiled by different compiles and stored in different files?

yes, that's the plan. But we are starting with only search for now

Comment (justinusmenzel) [reply scripts/compile-igs.js]: Maybe the ImplementationGuides interface could have a getName() method. That way the compile-igs script just needs to pass a list of ImplementationGuides to the IGCompiler and the compiler can use the getName() method to get the file name like so: `` const fileName = `${igs.getName()}.json` ``

**Comment by justinusmenzel** [comment]
The changes in the MR depend on two pending MRs (https://github.com/awslabs/fhir-works-on-aws-interface/pull/44 and https://github.com/awslabs/fhir-works-on-aws-search-es/pull/38). Do we need to wait for them to be merged before we can update package.json? Right now package.json points to `search-es` and `interface` versions without the changes that this MR depends on.

**Comment by carvantes** [comment]
> The changes in the MR depend on two pending MRs ([awslabs/fhir-works-on-aws-interface#44](https://github.com/awslabs/fhir-works-on-aws-interface/pull/44) and [awslabs/fhir-works-on-aws-search-es#38](https://github.com/awslabs/fhir-works-on-aws-search-es/pull/38)). Do we need to wait for them to be merged before we can update package.json? Right now package.json points to `search-es` and `interface` versions without the changes that this MR depends on.

Yes, those PRs need to be merged and new versions of the NPM packages released before we can update the `package.json` file on this PR. However, you can pull the code from those PRs and test locally

There are more detailed instructions on how to setup the multiple FWoA packages locally for development at: https://github.com/awslabs/fhir-works-on-aws-deployment/blob/mainline/DEVELOPMENT.md

File: scripts/compile-igs.js
Comment (carvantes) [reply scripts/compile-igs.js]: That's a good idea. Although I'm not making additional changes to the interface for the search-only release. I think that we could further explore this idea once we move to the next phase and start working on additional compilers.

**Review by carvantes — CHANGES_REQUESTED** [review]
I'm running the compile process locally and I have a couple of notes regarding `compile-igs.ts` :

- It should not fail if the `implementationGuides` folder does not exist. If it does not exist the script should simply be a no op since there's nothing to compile. Maybe just print a console message saying that no IGs were found.
- It should not fail if the `compiledImplementationGuides` folder does not exist. If it does not exist, the script should create it as part of writing the output file.

File: src/IGCompiler.test.ts
Comment (carvantes) [inline src/IGCompiler.test.ts]: Could you fix the eslint warnings? They don't fail the build but so far this package doesn't have any warnings and it'd be nice to keep it that way. For all these warnings just use arrow functions instead `() => {}`

File: src/IGCompiler.ts
Comment (carvantes) [inline src/IGCompiler.ts]: Can we change `collectResources` to return an array and omit the `resources` parameter? I think that creating an empty array on L90 shouldn't be necessary

File: src/IGCompiler.ts
Comment (carvantes) [inline src/IGCompiler.ts]: I think we should throw an error here. 

This condition will rarely happen, but it would mean that the IG cannot be processed. It'd be better to fail the compilation instead of skipping an entire IG silently.

File: src/IGCompiler.ts
Comment (carvantes) [inline src/IGCompiler.ts]: We should throw an error here. Same reason as my other comment regarding a missing `index.json` file

File: src/IGCompiler.ts
Comment (carvantes) [inline src/IGCompiler.ts]: can we refactor this into a separate private method?

I'm thinking of `validateDependencies(igs: IGInfo[]): void`. That method builds the `parentMap` and calls `depthFirst`

I think that makes it clearer that the intention is to validate. Also, I was a bit confused about `parentMap` at first but now I see that it is an auxiliary data structure only meant to be used alongside `depthFirst`. Having those 2 things inside of `validateDependencies` would make it clearer.

File: src/IGCompiler.ts
Comment (carvantes) [inline src/IGCompiler.ts]: I think we should add an `else` here and log a message like `Skipping hl7.fhir.r4.core since the base FHIR definitions are already included in fhir-works-on-aws`. It may be confusing for some users if we just skip it silently.

Also, I'm getting an error when trying to compile us core: `Exception:  Missing dependency hl7.fhir.r4.core@4.0.1 Error: Missing dependency hl7.fhir.r4.core@4.0.1`. This shouldn't happen since `hl7.fhir.r4.core` is already included in FWoA. We should ignore it on the dependency check.

Apologies if my previous comment about making an exception for `hl7.fhir.r4.core` was not clear about the dependencies part.

**Review by carvantes — CHANGES_REQUESTED** [review]

File: src/IGCompiler.ts
Comment (justinusmenzel) [reply src/IGCompiler.ts]: I can change the code to:
```
        const resources: any[] = [];
        for (const igInfo of await this.collectIGInfos(igsDir)) {
            resources.push(...(await this.collectResources(igInfo.path)));
        }
```
I don't see how I could get rid of creating the empty array in L90.

File: src/IGCompiler.ts
Comment (justinusmenzel) [reply src/IGCompiler.ts]: For testing it was convenient to point the compiler to a node_modules folder where some IGs were installed next to some other dependencies.

File: src/IGCompiler.ts
Comment (carvantes) [reply src/IGCompiler.ts]: Your proposed change looks good. 

you are right, an array is needed to collect all the resources. My bad, didn't notice that `collectResources` was being called within a loop.

**Comment by carvantes** [comment]
@justinusmenzel, I just merged this PR into a feature branch. There are still a few more things that need to be done for this feature before it can be merged back into `develop`

Thanks again for contributing!

**Pull request merged**
